### PR TITLE
feat(alertmanager-extra-infos): Adding extra labels or annotations the alert

### PR DESCRIPTION
**Pull request by Lowaiz**
Signed-off-by: Lyonel Martinez <lyonel.martinez@numberly.com>

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

/kind feature

**Any specific area of the project related to this PR?**

/area outputs

**What this PR does / why we need it**:

Now, we can add extra labels and extra annotations to the alerts send to the AlertManager Output.
Each label/annotation can have a name and a value defined by this format:
```
config:
    alertmanager:
        extralabels: "label1:value1, label2:value2"
        extraannotations: "annotation1:value1, annotation2:value2"
```

**Which issue(s) this PR fixes**:
None

**Special notes for your reviewer**:

We can use the same format to add extra labels with values to the Prometheus' extra-labels configuration and to the Loki one's.

### CODE REVIEW

File: outputs/alertmanager.go [hunk 1]
```diff
@@ -83,6 +84,35 @@ func newAlertmanagerPayload(falcopayload types.FalcoPayload, config *types.Confi
 	if config.Alertmanager.ExpiresAfter != 0 {
 		amPayload.EndsAt = falcopayload.Time.Add(time.Duration(config.Alertmanager.ExpiresAfter) * time.Second)
 	}
+	regProm, _ := regexp.Compile("^[a-zA-Z_:][a-zA-Z0-9_:]*$")
```
Comment (Issif) [inline outputs/alertmanager.go:87]: this should be compiled only once, at client creation, not for each event

File: outputs/alertmanager.go [hunk 2]
```diff
@@ -83,6 +84,35 @@ func newAlertmanagerPayload(falcopayload types.FalcoPayload, config *types.Confi
...
+	for _, label := range config.Alertmanager.ExtraLabelsList {
+		data := strings.SplitN(label, "=", 2)
+		labelName := data[0]
+		labelName = strings.ReplaceAll(strings.TrimSpace(labelName), ".", "_")
+		labelValue := ""
+		if len(data) == 2 {
+			labelValue = data[1]
+		}
+		if !regProm.MatchString(labelName) {
+			log.Printf("[ERROR] : Extra field '%v' is not a valid prometheus label", label)
```
Comment (Issif) [inline outputs/alertmanager.go:97]:
```suggestion
			log.Printf("[ERROR] : AlertManager - Extra field '%v' is not a valid prometheus label", label)
```

File: outputs/alertmanager.go [hunk 3]
```diff
@@ -83,6 +84,35 @@ func newAlertmanagerPayload(falcopayload types.FalcoPayload, config *types.Confi
...
+	for _, annotation := range config.Alertmanager.ExtraAnnotationsList {
+		data := strings.SplitN(annotation, "=", 2)
+		annotationName := data[0]
+		annotationName = strings.ReplaceAll(strings.TrimSpace(annotationName), ".", "_")
+		annotationValue := ""
+		if len(data) == 2 {
+			annotationValue = data[1]
+		}
+		if !regProm.MatchString(annotationName) {
+			log.Printf("[ERROR] : Extra field '%v' is not a valid prometheus annotation", annotation)
```
Comment (Issif) [inline outputs/alertmanager.go:111]:
```suggestion
			log.Printf("[ERROR] : AlertManager - Extra field '%v' is not a valid prometheus annotation", annotation)
```

File: types/types.go [hunk 4]
```diff
@@ -167,12 +167,16 @@ type DiscordOutputConfig struct {
...
-	Endpoint        string
-	ExpiresAfter    int
+	HostPort             string
+	MinimumPriority      string
+	CheckCert            bool
+	MutualTLS            bool
+	Endpoint             string
+	ExpiresAfter         int
+	ExtraLabels          string
+	ExtraLabelsList      []string
```
Comment (Issif) [inline types/types.go:177]: Why are you creating a slice and not a map[string]string?

**Review by Issif — CHANGES_REQUESTED** [review]
Can you also update the README (both section, example file + env vars) + the config_example.yaml, please

File: types/types.go (hunk 4 above)
Comment (Issif) [reply types/types.go:177]: the parsing should be done at the client creation too

Comment (Lowaiz) [reply types/types.go:177]: I wanted to have constancy with Prometheus config and Loki one's.
I also didn't see there was already everything to parse the map from env vars.

Comment (Lowaiz) [reply types/types.go]: I guess it's cleaner now :D

File: outputs/alertmanager.go [hunk 5]
```diff
@@ -83,6 +94,20 @@ func newAlertmanagerPayload(falcopayload types.FalcoPayload, config *types.Confi
 	if config.Alertmanager.ExpiresAfter != 0 {
 		amPayload.EndsAt = falcopayload.Time.Add(time.Duration(config.Alertmanager.ExpiresAfter) * time.Second)
 	}
+	for label, value := range config.Alertmanager.ExtraLabels {
+		if !regProm.MatchString(label) {
```
Comment (Issif) [inline outputs/alertmanager.go:98]: the extra labels/annotations are static, we should check if they match the regexp in config.go, when we map the values, no need to check at each event.

**Review by Issif — APPROVED** [review]
:+1:

**Pull request merged**